fuse-utils/man/createhdf.1
.\" -*- nroff -*-
.\"
.\" createhdf.1: createhdf man page
.\" Copyright (c) 2004 Philip Kendall
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write to the Free Software Foundation, Inc.,
.\" 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
.\"
.\" Author contact information:
.\"
.\" E-mail: philip-fuse@shadowmagic.org.uk
.\"
.\"
.TH createhdf 1 "21st August, 2016" "Version 1.2.2" "Emulators"
.\"
.\"------------------------------------------------------------------
.\"
.SH NAME
createhdf \(em Create a blank IDE hard disk image in .hdf format
.\"
.\"------------------------------------------------------------------
.\"
.SH SYNOPSIS
.B createhdf
[ \-c ] [ \-s ]
.RI "[\-v " version ]
.I cylinders heads sectors file
.P
.\"
.\"------------------------------------------------------------------
.\"
.SH DESCRIPTION
createhdf creates a blank image of an IDE hard disk in .hdf format for
use with Sinclair ZX Spectrum emulators.
.\"
.\"------------------------------------------------------------------
.\"
.SH OPTIONS
.TP
.I \-c
specifies that the .hdf image should be created in `compact' mode,
where only the low byte of every word is stored in the image. This is
useful for saving space if the hard disk image is for use with an
interface which accesses only the low byte of every word.
.TP
.I \-s
specifies that the .hdf image should
.B not
be created as a sparse file. Normally,
.I createhdf
will attempt to create the hard disk image as a `sparse file', where
the operating system knows that the file consists mostly of zero bytes
and fills those in without using any space in the file. This option
disables this. (There's nothing `magic' about this: if data is written
to the file, it will take up space just like any other data; it's only
empty space which is synthesised by the operating system).
.TP
.I \-v version
specifies the version of .hdf image to be created. Valid values are
.RI ` 1.0 '
for a version 1.0 file and
.RI ` 1.1 '
for a version 1.1 file. Defaults to creating version 1.1 files.
.TP
.IR \-h ", " \-\-help
give brief usage help, listing available options.
.TP
.IR \-V ", " \-\-version
output version information.
.TP
.I cylinders
specifies the number of cylinders in the image.
.TP
.I heads
specifies the number of heads in the image.
.TP
.I sectors
specifies the number of sectors in the image.
.TP
.I file
specifies the file to which the image should be written.
.\"
.\"------------------------------------------------------------------
.\"
.SH BUGS
None known.
.\"
.\"------------------------------------------------------------------
.\"
.SH SEE ALSO
.IR fuse "(1),"
.IR fuse\-utils "(1),"
.IR libspectrum "(3)"
.PP
The .hdf format specification, at
.br
.IR "http://www.zxshed.co.uk/sinclairfaq/index.php5?title=HDF_format" .
.PP
The comp.sys.sinclair Spectrum FAQ, at
.br
.IR "http://www.worldofspectrum.org/faq/index.html" .
.\"
.\"------------------------------------------------------------------
.\"
.SH AUTHOR
Philip Kendall (philip\-fuse@shadowmagic.org.uk).


fuse/ChangeLog
2016-07-17  Philip Kendall  <philip-fuse@shadowmagic.org.uk>

        * Fuse 1.2.1 released.

        * Emulation core improvements:
          * Fix bugs when the detect loaders feature is being used (thanks, Lee
            Tonks and windale) (Fredrick Meunier).

        * Debugger improvements:
          * Remove the need for "%" when accessing system variables (Philip
            Kendall).
          * Add Z80 registers as debugger variables (Philip Kendall).
          * Expose last byte written to the ULA, tstates since interrupt,
            primary and secondary memory control ports as debugger system
            variables (Philip Kendall).
          * Make breakpoints on events honour lifetime (Sergio Baldoví).
          * Extend breakpoints on paging events to more peripherals: Beta 128,
            +D, Didaktik 80, DISCiPLE, Opus Discovery and SpeccyBoot (Sergio
            Baldoví).
          * Split +D memory sources into RAM and ROM sections (Sergio Baldoví).
          * Coalesce +D and DISCiPLE RAM pages so they show as 8K pages (Sergio
            Baldoví).

        * Miscellaneous improvements:
          * Update GNOME .desktop file (Alberto Garcia).
          * Add an emulator module startup manager to automatically handle
            dependency issues (Philip Kendall).
          * Fix crash on widget UIs when hitting the close icon on the title bar
            several times (Sergio Baldoví).

2016-06-06  Philip Kendall  <philip-fuse@shadowmagic.org.uk>

        * Fuse 1.2.0 released.

        * New features:
          * Add Currah µSource emulation (Stuart Brady).
          * Add Didaktik 80/40 emulation (Gergely Szasz).
          * Capture BASIC video functions to SVG file (Stefano Bodrato).
          * Support bash completion (Sergio Baldoví).

        * Emulation core improvements:
          * Allow continuing RZX recordings if there is a final snapshot in the
            RZX (Sergio Baldoví).
          * Fix the prune function on RZX rollback (Sergio Baldoví).
          * Use SZX format for the initial snapshot in RZX files (Sergio
            Baldoví).
          * Fix loading of EDSK files with Sector Offset block (Sergio Baldoví).
          * Migrate disk "index event" handling to the FDD layer and have the
            FDC layer use it for their STATUS registers (Gergely Szasz).
          * Implement WD2797 emulation (Gergely Szasz).
          * Centralise the "Disk icon" update code to the FDD layer (Gergely
            Szasz).
          * Fix disk image corruption after saving UDI files (Sergio Baldoví).
          * Check ready status after loading a disk into floppy disk drive
            (thanks, John Elliott) (Sergio Baldoví).
          * Fix overlapped SEEK commands (Sergio Baldoví).
          * Fix length of data returned by READ_DIAG (thanks, Fredrick Meunier)
            (Sergio Baldoví).
          * Fix writing .td0 format disk files (Sergio Baldoví).
          * Fix speech in Cobra's Arc - Medium Case.tzx when loaded with tape
            traps enabled (thanks, zx81 and Sergio Baldoví) (Fredrick Meunier).
          * Skip tape traps if VERIFY is requested (UB880D).
          * Fix loading sound with some custom loaders (Fredrick Meunier).
          * Check if data blocks are headers for handling PZX files (Fredrick
            Meunier).
          * Correct display of pulses in PZX pulse block for tape browser
            (Fredrick Meunier).
          * Set AF, AF' and SP to 0xffff on reset (Stuart Brady).
          * Leave most registers unchanged on a soft reset (Stuart Brady).
          * Emulate interrupt and NMI timings more precisely (Stuart Brady).
          * Emulate NMOS and CMOS Z80 variants (Stuart Brady).
          * Switch to 2KB page size (Stuart Brady).
          * Fix inaccurate output when AY envelopes are used (Matthew Westcott
            and Fredrick Meunier).
          * Fix some peripherals activation when loading snapshots (Sergio
            Baldoví).

        * Machine specific improvements:
          * The address range 0x4000 - 0x7FFF is contended on the TS2068 in the
            home, Dock and Exrom banks (thanks, Richard Atkinson). It is assumed
            that this is the same for other Timex models (Fredrick Meunier).
          * Clear all Opus RAM on hard reset (Stuart Brady).
          * Ensure the ZX Printer does not require a hard reset to enable
            (thanks, RMartins) (Fredrick Meunier).
          * Fix timing of events when emulating Scorpion (Stuart Brady).
          * Ensure we have successfully selected a Pentagon or a Timex machine
            before inserting their media (Fredrick Meunier).
          * Prefer Scorpion to Pentagon when loading SCL/TRD disks for better
            timing compatibility (thanks, windale) (Sergio Baldoví).
          * Fix Beta 128 type II commands (thanks, windale and BogDan Vatra)
            (Sergio Baldoví).
          * Don't enable the Beta 128 interface when loading a snapshot on a
            machine with Beta built-in (thanks, windale and BogDan Vatra)
            (Fredrick Meunier).
          * Lock port +3 1FFDh if paging is disabled (Brian Ruthven and Fredrick
            Meunier).
          * Spectranet: fix segfault in error handling when setting SO_REUSEADDR
            (Stuart Brady).
          * Update SE ROMs to v4.07 (thanks, Andrew Owen) (Sergio Baldoví).

        * Debugger improvements:
          * Fix disassembly of LD (HL), LD (IX) and LD (IY) (BogDan Vatra).
          * Add I and R register setting and getting to the debugger (Sergio
            Baldoví).
          * Show the status of the halted flag in the debugger (Stuart Brady).
          * Fix time breakpoints later than a frame in the future (Sergio
            Baldoví).
          * Timex EXROM and Dock text was truncated in the debugger UI (thanks,
            Andrew Owen and Sergio Baldoví) (Fredrick Meunier).
          * Win32: relocate halted flag and set monospaced font (Sergio

fuse-utils/createhdf.c
/* createhdf.c: Create an empty .hdf file
   Copyright (c) 2004-2006 Philip Kendall
   Copyright (c) 2015 Sergio Baldovi

   $Id$

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License along
   with this program; if not, write to the Free Software Foundation, Inc.,
   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

   Author contact information:

   Philip Kendall <philip-fuse@shadowmagic.org.uk>

*/

#include <config.h>

#include <errno.h>
#include <fcntl.h>
#include <getopt.h>
#include <stdio.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>

#include "compat.h"
#include "ide.h"

#define PROGRAM_NAME "createhdf"

const char *progname;

#define CHUNK_LENGTH 1 << 20

static void
show_version( void )
{
  printf(
    PROGRAM_NAME " (" PACKAGE ") " PACKAGE_VERSION "\n"
    "Copyright (c) 2004-2006 Philip Kendall\n"
    "License GPLv2+: GNU GPL version 2 or later "
    "<http://gnu.org/licenses/gpl.html>\n"
    "This is free software: you are free to change and redistribute it.\n"
    "There is NO WARRANTY, to the extent permitted by law.\n" );
}

static void
show_help( void )
{
  printf(
    "Usage: %s [OPTION]... <cylinders> <heads> <sectors> <file>\n"
    "Creates a blank image of an IDE hard disk in .hdf format.\n"
    "\n"
    "Options:\n"
    "  -c             Create .hdf image in compact mode, where only the low byte\n"
    "                   of every word is stored in the image.\n"
    "  -s             Do not create .hdf image as a sparse file, i.e., empty\n"
    "                   space synthesised by the operating system.\n"
    "  -v <version>   Specifies the version of .hdf image to be created (values\n"
    "                   1.0 or 1.1). Defaults to creating version 1.1 files.\n"
    "  -h, --help     Display this help and exit.\n"
    "  -V, --version  Output version information and exit.\n"
    "\n"
    "Non-option arguments:\n"
    "  cylinders      Specifies the number of cylinders in the image.\n"
    "  heads          Specifies the number of heads in the image.\n"
    "  sectors        Specifies the number of sectors in the image.\n"
    "  file           Specifies the file to which the image should be written.\n"
    "\n"
    "Report %s bugs to <%s>\n"
    "%s home page: <%s>\n"
    "For complete documentation, see the manual page of %s.\n",
    progname,
    PROGRAM_NAME, PACKAGE_BUGREPORT, PACKAGE_NAME, PACKAGE_URL, PROGRAM_NAME
  );
}

static int
parse_options( int argc, char **argv, size_t *cylinders, size_t *heads,
	       size_t *sectors, int *compact, int *sparse,
	       const char **filename, enum hdf_version_t *version )
{
  int c;
  int error = 0;

  struct option long_options[] = {
    { "help", 0, NULL, 'h' },
    { "version", 0, NULL, 'V' },
    { 0, 0, 0, 0 }
  };

  while( ( c = getopt_long( argc, argv, "csv:hV", long_options,
                            NULL ) ) != -1 ) {

    switch( c ) {

    case 'c': *compact = 1; break;
    case 's': *sparse = 0; break;
    case 'v':
      if( strcasecmp( optarg, "1.0" ) == 0 ) {
	*version = HDF_VERSION_10;
      } else if( strcasecmp( optarg, "1.1" ) == 0 ) {
	*version = HDF_VERSION_11;
      } else {
        error = 1;
        fprintf( stderr, "%s: version not supported\n", progname );
      }
      break;

    case 'h':
      show_help();
      exit( 0 );

    case 'V':
      show_version();
      exit( 0 );

    case '?':
      /* getopt prints an error message to stderr */
      error = 1;
      break;

    default:
      error = 1;
      fprintf( stderr, "%s: unknown option `%c'\n", progname, (char) c );
      break;
    }
  }

  if( error ) return error;

  if( argc - optind != 4 ) {
    fprintf( stderr,
	     "%s: usage: %s [-c] [-s] [-v<version>] <cylinders> <heads> <sectors> <hdf>\n",
	     progname, progname );
    return 1;
  }

  *cylinders = atoi( argv[ optind++ ] );
  *heads     = atoi( argv[ optind++ ] );
  *sectors   = atoi( argv[ optind++ ] );
  *filename  = argv[ optind++ ];

  return 0;
}

static int
write_header( int fd, size_t cylinders, size_t heads, size_t sectors,
	      int compact, const char *filename, enum hdf_version_t version )
{
  char *hdf_header, *identity;
  size_t bytes, data_offset;

  data_offset = hdf_data_offset( version );

  hdf_header = calloc( 1, data_offset );
  if( !hdf_header ) {
    fprintf( stderr, "%s: out of memory at line %d\n", progname, __LINE__ );
    return 1;
  }

  memcpy( &hdf_header[ HDF_SIGNATURE_OFFSET ], HDF_SIGNATURE,
	  HDF_SIGNATURE_LENGTH );
  hdf_header[ HDF_VERSION_OFFSET ] = hdf_version( version );
  hdf_header[ HDF_COMPACT_OFFSET ] = compact;
  hdf_header[ HDF_DATA_OFFSET_OFFSET     ] = ( data_offset      ) & 0xff;
  hdf_header[ HDF_DATA_OFFSET_OFFSET + 1 ] = ( data_offset >> 8 ) & 0xff;

  identity = &hdf_header[ HDF_IDENTITY_OFFSET ];

  identity[ IDENTITY_CYLINDERS_OFFSET     ] = ( cylinders      ) & 0xff;
  identity[ IDENTITY_CYLINDERS_OFFSET + 1 ] = ( cylinders >> 8 ) & 0xff;

  identity[ IDENTITY_HEADS_OFFSET         ] = ( heads          ) & 0xff;
  identity[ IDENTITY_HEADS_OFFSET     + 1 ] = ( heads     >> 8 ) & 0xff;

  identity[ IDENTITY_SECTORS_OFFSET       ] = ( sectors        ) & 0xff;
  identity[ IDENTITY_SECTORS_OFFSET   + 1 ] = ( sectors   >> 8 ) & 0xff;

  bytes = write( fd, hdf_header, data_offset );
  if( bytes != data_offset ) {
    fprintf( stderr,
	     "%s: could write only %lu header bytes out of %lu to '%s'\n",
	     progname, (unsigned long)bytes, (unsigned long)data_offset,
	     filename );
    free( hdf_header );
    return 1;
  }

  free( hdf_header );

  return 0;
}

static int
write_data( int fd, size_t cylinders, size_t heads, size_t sectors,
	    int compact, int sparse, const char *filename,
	    enum hdf_version_t version )
{
  size_t data_size;

  data_size = cylinders * heads * sectors * ( compact ? 0x100 : 0x200 );

  if( sparse ) {

    if( ftruncate( fd, data_size + hdf_data_offset( version ) ) ) {
      fprintf( stderr, "%s: error truncating '%s': %s\n", progname, filename,
	       strerror( errno ) );
      return 1;
    }

  } else {

    char buffer[ CHUNK_LENGTH ];
    size_t total_written;

    memset( buffer, 0, CHUNK_LENGTH );
    total_written = 0;

    while( data_size ) {

      size_t bytes_to_write, bytes_written;

      bytes_to_write = data_size > CHUNK_LENGTH ? CHUNK_LENGTH : data_size;

      bytes_written = write( fd, buffer, bytes_to_write );
      if( bytes_written != bytes_to_write ) {
	fprintf( stderr, "%s: could write only %lu bytes out of %lu to '%s'",
		 progname, (unsigned long)total_written + bytes_written,
		 (unsigned long)data_size, filename );
	return 1;
      }

      total_written += bytes_to_write;
      data_size -= bytes_to_write;
    }
  }

  return 0;
}

int
main( int argc, char **argv )
{
  size_t cylinders, heads, sectors;
  int compact, sparse;
  enum hdf_version_t version;
  const char *filename;
  int fd;
  int error;

  progname = argv[0];

  compact = 0;
  sparse = 1;
  version = HDF_VERSION_11;
  error = parse_options( argc, argv, &cylinders, &heads, &sectors, &compact,
			 &sparse, &filename, &version );
  if( error ) {
    fprintf( stderr, "Try `%s --help' for more information.\n", progname );
    return error;
  }

  fd = open( filename, O_WRONLY | O_CREAT | O_TRUNC | O_BINARY,
			S_IRUSR | S_IWUSR |
			S_IRGRP | S_IWGRP |
			S_IROTH | S_IWOTH   );
  if( fd == -1 ) {
    fprintf( stderr, "%s: error opening '%s': %s\n", progname, filename,
	     strerror( errno ) );
    return 1;
  }

  error = write_header( fd, cylinders, heads, sectors, compact, filename,
			version );
  if( error ) return error;

  error = write_data( fd, cylinders, heads, sectors, compact, sparse,
		      filename, version );
  if( error ) return error;

  return 0;
}
